elastic/kibana
# [ResponseOps] resolve conflicts when updating alert docs after rule execution

resolves: #158403

When conflicts are detected while updating alert docs after a rule runs, we'll try to resolve the conflict by `mget()`'ing the alert documents again, to get the updated OCC info `_seq_no` and `_primary_term`.  We'll also get the current versions of "ad-hoc" updated fields (which caused the conflict), like workflow status, case assignments, etc.  And then attempt to update the alert doc again, with that info, which should get it back up-to-date.

Note that the rule registry was not touched here.  During this PR's development, I added the retry support to it, but then my function tests were failing because there were never any conflicts happening.  Turns out rule registry mget's the alerts before it updates them, to get the latest values.  So they won't need this fix.

It's also not clear to me if this can be exercised in serverless, since it requires the use of an alerting framework based AaD implementation AND the ability to ad-hoc update alerts.  I think this can only be done with Elasticsearch Query and Index Threshold, and only when used in metrics scope, so it will show up in the metrics UX, which is where you can add the alerts to the case.

## manual testing

It's hard!  I've seen the conflict messages before, but it's quite difficult to get them to go off whenever you want.  The basic idea is to get a rule that uses alerting framework AAD (not rule registry, which is not affected the same way with conflicts (they mget alerts right before updating them), set it to run on a `1s` interval, and probably also configure TM to run a `1s` interval, via the following configs:

```
xpack.alerting.rules.minimumScheduleInterval.value: "1s"
xpack.task_manager.poll_interval: 1000
```

You want to get the rule to execute often and generate a lot of alerts, and run for as long as possible.  Then while it's running, add the generated alerts to cases.  Here's the EQ rule definition I used:

![image](https://github.com/elastic/kibana/assets/25117/56c69d50-a76c-48d4-9a45-665a0008b248)

I selected the alerts from the o11y alerts page, since you can't add alerts to cases from the stack page.  Hmm.  :-). Sort the alert list by low-high duration, so the newest alerts will be at the top.  Refresh, select all the rules (set page to show 100), then add to case from the `...` menu.  If you force a conflict, you should see something like this in the Kibana logs:

```
[ERROR] [plugins.alerting] Error writing alerts: 168 successful, 100 conflicts, 0 errors:
[INFO ] [plugins.alerting] Retrying bulk update of 100 conflicted alerts
[INFO ] [plugins.alerting] Retried bulk update of 100 conflicted alerts succeeded
```
